Use io_uring in MSRIO when batched msr-safe is not available

dannosliwcd opened the pull request:
This change uses io_uring when batched msr-safe is not available. The change adds an optional build dependency on liburing to interact with io_uring, falling back to regular IO if liburing is not present or if liburing does not report that it supports the capabilities we need to use.

Resolves #2781

dannosliwcd committed:
Add an io_uring wrapper class

- Add an interface that lets a user push reads or writes to a work queue, then
  submit the queue in bulk.
- Add implementations that use io_uring if available, or use a series of pread
  and pwrite operations otherwise.

dannosliwcd committed:
Use io_uring in MSRIO when batched msr-safe is not available

- Resolves #2781

dannosliwcd commented on service/src/IOUringImp.hpp:
```diff
@@ -0,0 +1,53 @@
+#ifndef IOURINGIMP_HPP_INCLUDE
+#define IOURINGIMP_HPP_INCLUDE
+
+#define _POSIX_C_SOURCE 200809L
```
Looks like maybe this needs to be _DEFAULT_SOURCE instead, for `loff_t`.

cmcantalupo replied:
I think we define _POSIX_C_SOURCE in configure.ac for all compilation units.  We may have to be careful about how these settings interact.

dannosliwcd replied:
From https://man7.org/linux/man-pages/man7/feature_test_macros.7.html, I think `_DEFAULT_SOURCE` is similar to `-D_BSD_SOURCE -D_SVID_SOURCE -D_POSIX_C_SOURCE=200809`. Should I switch the configure option to use that instead?

cmcantalupo replied:
We define it in Makefile.am:

https://github.com/geopm/geopm/blob/dev/service/Makefile.am#L7-L9

I wonder if it would be best to just put

```
#define _BSD_SOURCE
#define _SVID_SOURCE
```
in the IOUringImp.hpp

and rely on the _POSIX_C_SOURCE that gets propagated through AM_CPPFLAGS

cmcantalupo commented on service/configure.ac:
```diff
@@ -195,6 +210,25 @@ if test "x$with_levelzero" != x; then
   EXTRA_LDFLAGS="$EXTRA_LDFLAGS -L$with_levelzero/lib64"
 fi
 
+AC_ARG_WITH([liburing], [AS_HELP_STRING([--with-liburing=PATH],
```
Need to add

```
BuildRequires: liburing-devel
```

to the geopm-service.spec.in file.  Looks like this is the right name for both SLES and RHEL rpm naming schemes.

dannosliwcd replied:
Also add the `--enable-io-uring` config option and `Requires: liburing` to that specfile.

cmcantalupo commented on service/src/IOUring.cpp:
```diff
@@ -0,0 +1,44 @@
+    }
+#endif
+
+    std::shared_ptr<IOUring> make_io_uring(unsigned entries)
```
I think a unique_ptr return type is preferred here unless the factory is intended to retain ownership.  The unique_ptr return value can be assigned to a shared_ptr.

cmcantalupo commented on service/src/IOUring.hpp:
```diff
@@ -0,0 +1,67 @@
+    /// @brief Create an object that supports an io_uring-like interface. The
+    ///        created object uses io_uring if supported, otherwise uses
+    ///        individual read/write operations.
+    std::shared_ptr<IOUring> make_io_uring(unsigned entries);
```
see comment above...probably should have put it here.  Also please make this a static member of the IOUring class.  I've been trying to follow this pattern with newer classes.  Typical use case would be:

```
    std::shared_ptr<IOUring> shared_uring_ptr = IOUring::make_unique();
    std::unique_ptr<IOUring> unique_uring_ptr = IOUring::make_unique();
```

cmcantalupo commented on service/src/IOUringFallback.cpp:
```diff
@@ -0,0 +1,61 @@
+
+    void IOUringFallback::submit()
+    {
+        for (const auto &operation : m_operations) {
+            errno = 0;
+            int ret = operation.second();
+            if (ret < 0) {
+                ret = -errno;
+            }
```
```suggestion
        errno = 0;
        for (const auto &operation : m_operations) {
            int ret = operation.second();
            if (ret < 0) {
                ret = -errno;
            }
            errno = 0;

```

cmcantalupo replied:
Make sure to leave errno zero when the loop completes.  Nice use of bind though! Very clean.

dannosliwcd commented on service/src/IOUringImp.cpp:
```diff
@@ -0,0 +1,118 @@
+                               unsigned nbytes, off_t offset)
+    {
+        auto sqe = get_sqe_or_throw();
+        set_sqe_return_destination(sqe, ret);
+        // Available since Linux 5.10
+        io_uring_prep_read(sqe, fd, buf, nbytes, offset);
```
Swap the order of these. Also make sure the test cases cover the user_data field usage.

dannosliwcd replied:
Looks like the tests already do cover that field (i.e., they assert if the retval is not tracked as expected). I guess the version of liburing installed on my test system doesn't clear the user_data field in prep_read. Anyways, the order still needs to be swapped so that we modify user_data _after_ the prep_... functions are called.

dannosliwcd commented:
Rebased on dev so CI isn't blocked on merge conflicts. Did not squash any commits in the rebase.

dannosliwcd commented on service/configure.ac:
```diff
@@ -133,6 +133,21 @@ fi
 [enable_cnl_iogroup="1"]
 )
 
+AC_ARG_ENABLE([io-uring],
```
Swap this around to enable by default, with a disable option.

cmcantalupo commented on service/geopm-service.spec.in:
```diff
@@ -107,6 +107,12 @@ Group: System Environment/Libraries
 Group: System/Libraries
 %endif
 
+# Optional dependency: Remove --enable-io-uring from the ./configure step. 
+# Lets GEOPM batch its IO operations to reduce syscall overhead in IOGroups
+# with a lot of reads/writes per GEOPM batch operation.
+BuildRequires: liburing-devel
+Requires: liburing
+
```
```suggestion
# Lets GEOPM batch its IO operations to reduce syscall overhead in IOGroups
# with a lot of reads/writes per GEOPM batch operation.
BuildRequires: liburing-devel

```

cmcantalupo replied:
The requires will be automatically be picked up by rpmbuild when the elf header is inspected.

cmcantalupo commented on service/geopm-service.spec.in:
```diff
@@ -163,6 +169,7 @@ CFLAGS= CXXFLAGS= CC=gcc CXX=g++ \
             --mandir=%{_mandir} --docdir=%{docdir} \
             --with-python=%{python_bin} \
             %{?level_zero_option} \
+            --enable-io-uring \
```
```suggestion
```

cmcantalupo reviewed (commented):
small suggested change to the spec file now that uring is enabled by default.  Will have to add the ubuntu equivalent to liburing-devel to the .github/build.yml file

dannosliwcd commented on .github/workflows/build.yml:
```diff
@@ -26,13 +26,13 @@ jobs:
     # Run linters and other low-cost checks that do not depend on being in a
     # particular configuration of the build matrix.
-    runs-on: ubuntu-20.04
+    runs-on: ubuntu-22.04
```
Run this by @bgeltz

dannosliwcd replied:
Handled in #2825

cmcantalupo commented on service/src/IOUringImp.hpp:
```diff
@@ -0,0 +1,55 @@
+            static bool is_supported();
+
+            /// @brief Create an IO uring with queues of a given size.
+            /// @param entries  Maximum number of queue operations to contain
+            ///                 within a single batch submission.
```
Add this parameter definition to the static IOUring::make_unique() dox, alternately, you can just call:

```
std::unique_ptr<IOUring> result = std::make_unique<IOUringImp>(entries);
```

in `IOUring::make_unique()`, and then `IOUringImp::make_unique()` can go away, but the implementation you have allows the `IOuring.cpp` file to avoid including `IOuringImp.hpp`.  It does have to include the fallback header though.

cmcantalupo commented on .github/workflows/build.yml:
```diff
@@ -66,7 +66,7 @@ jobs:
     - uses: actions/checkout@v2
     - name: install system dependencies
-      run: sudo apt-get update && sudo apt-get install libelf-dev mpich libmpich-dev libomp-11-dev libsystemd-dev python3-gi python3-yaml
+      run: sudo apt-get update && sudo apt-get install libelf-dev mpich libmpich-dev libomp-11-dev libsystemd-dev liburing-dev python3-gi python3-yaml
```
please add to

https://github.com/geopm/geopm/blob/dev/service/debian/control#L5-L18

as well!

cmcantalupo committed:
Opt out of liburing support for SLES versions older than 15 SP3

- Change library check to look for symbol missing from SLES 15 SP2
  distribution.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>
